opm/simulators/wells/ParallelWellInfo.cpp

```diff
@@ -0,0 +1,325 @@
+/*
+  Copyright 2020 OPM-OP AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <config.h>
+#include <opm/simulators/wells/ParallelWellInfo.hpp>
+#include <opm/common/ErrorMacros.hpp>
+
+namespace Dune
+{
+#if HAVE_MPI
+template<>
+struct CommPolicy<double*>
+{
+    using Type = double*;
+    using IndexedType = double;
+    using IndexedTypeFlag = Dune::SizeOne;
+    static const void* getAddress(const double*& v, int index)
+    {
+        return v + index;
+    }
+    static int getSize(const double*&, int)
+    {
+        return 1;
+    }
+};
+#endif
+}
+
+namespace Opm
+{
+CommunicateAbove::CommunicateAbove([[maybe_unused]] const Communication& comm)
+#if HAVE_MPI
+    : comm_(comm), interface_(comm_)
+#endif
+{}
+
+void CommunicateAbove::clear()
+{
+#if HAVE_MPI
+    above_indices_ = {};
+    current_indices_ = {};
+    interface_.free();
+    communicator_.free();
+#endif
+    count_ = 0;
+}
+
+void CommunicateAbove::beginReset()
+{
+    clear();
+#if HAVE_MPI
+    if (comm_.size() > 1)
+    {
+        current_indices_.beginResize();
+        above_indices_.beginResize();
+    }
+#endif
+}
+
+void CommunicateAbove::endReset()
+{
+#if HAVE_MPI
+    if (comm_.size() > 1)
+    {
+        above_indices_.endResize();
+        current_indices_.endResize();
+        remote_indices_.setIndexSets(current_indices_, above_indices_, comm_);
+        remote_indices_.setIncludeSelf(true);
+        remote_indices_.rebuild<true>();
+        using FromSet = Dune::EnumItem<Attribute,owner>;
+        using ToSet = Dune::AllSet<Attribute>;
+        interface_.build(remote_indices_, FromSet(), ToSet());
+        communicator_.build<double*>(interface_);
+    }
+#endif
+}
+
+std::vector<double> CommunicateAbove::communicate(double first_above,
```

Comment:
I have some trouble understanding what is supposed to happen here. Could you give a simple example?

Reply:
Thanks, that was clarifying, I now understand the purpose of the class. I think that it is correct for the single-process case. I have not checked the actually communicating case, but will assume that is ok.